src/vsr/checksum.zig

```diff
@@ -1,132 +1,89 @@
+//! This file implements vsr.checksum. TigerBeetle uses this checksum to:
+//!
+//! - detect bitrot in data on disk,
+//! - validate network messages before casting raw bytes to an `extern struct` type,
+//! - hash-chain prepares and client requests to have strong consistency and ordering guarantees.
+//!
+//! As this checksum is stored on disk, it is set in stone and impossible to change.
+//!
+//! We need this checksum to be fast (it's in all our hotpaths) and strong (it's our ultimate line
+//! of defense against storage failures and some classes of software bugs).
+//!
+//! Our checksum of choice is based on Aegis:
+//!
+//! <https://datatracker.ietf.org/doc/draft-irtf-cfrg-aegis-aead/>
+//!
+//! We use the implementation from the Zig standard library, but here's the overall overview of the
+//! thing works:
+//!
+//! - AES-block is a widely supported in hardware symmetric encryption primitive (`vaesenc`,
+//!   `vaesdec` instructions). Hardware acceleration is what provides speed.
+//! - Aegis is an modern Authenticated Encryption with Associated Data (AEAD) scheme based on
+//!   AES-block.
+//! - In AEAD, the user provides, a key, a nonce, a secret message, and associated data, and gets
+//!   a ciphertext and an authentication tag back. Associated data is expected to be sent as plain
+//!   text (eg, it could be routing information). The tag authenticates _both_ the secret message
+//!   and associated data.
+//! - AEAD can be specialized to be a MAC by using an empty secret message and zero nonce. NB:
+//!   in mac mode, message to sign is treated as AD, not as a secret message.
+//! - A MAC can further be specialized to be a checksum by setting the secret key to zero.
+//!   And that's what we do here!
+
 const std = @import("std");
 const builtin = @import("builtin");
 const mem = std.mem;
 const testing = std.testing;
+const assert = std.debug.assert;
 
-// TODO(king): Replace with std.crypto.aead.aegis.Aegis128LMac_128 once Zig is updated.
-const Aegis128 = struct {
```

Comment:
Side note: wondering if we'll need to bring this back for checksum() to work at comptime (thinking `@inComptime()` switch for `aes.Block`)